feat: add CSV import with streaming progress and row-level validation

alserious opened the pull request:
Adds opt-in CSV import to the model list page. Users upload a UTF-8 CSV file; the server validates each row (WTForms + foreign keys), optionally persists valid rows in a single transaction with per-row savepoints, and streams progress/results as NDJSON.

Import is disabled by default (can_import = False) and gated by check_can_import(request).

### ModelView configuration

| Option | Default | Purpose |
|--------|---------|---------|
| `can_import` | `False` | Enable import UI + endpoint |
| `column_import_list` | list page columns / PK | Columns expected in CSV |
| `column_import_exclude_list` | — | Exclude columns from import |
| `max_import_file_size` | 5 MiB | Max upload size |
| `import_max_rows` | `0` (unlimited) | Max data rows per file |
| `max_reported_missed_rows` | `100` | Cap on detailed error payload |

<img width="397" height="292" alt="image" src="https://github.com/user-attachments/assets/462e9f61-23ed-42c4-84d0-529acd42d393" />

alserious committed:
init import feature

alserious committed:
fix lint

alserious committed:
fix tests

alserious committed:
remove bom byte if exist

alserious committed:
filter import columns

alserious committed:
fix lint

alserious committed:
fix mypy

alserious committed:
fix bom byte and import columns

alserious committed:
remove unused

alserious commented on sqladmin/application.py:
```diff
@@ -629,6 +641,58 @@ async def export(self, request: Request) -> Response:
+                for row in data:
+                    n_row = []
+                    for value in row.getlist(relation_name):
+                        if value == str(relation_obj):
```
Yes, it seems like this is the only way, as long as we don't use a primary key list. If we remove the relationship import, it won't be a problem.

mmzeynalli replied:
@alserious this is nice feature. Do you plan to fix it so we can merge it?

alserious replied:
@mmzeynalli removed the import of relations, simple import works

alserious committed:
remove import relation

alserious committed:
remove relation 2 part

alserious committed:
fix tests

alserious committed:
fix from rebase

alserious committed:
fix

alserious committed:
fix

mmzeynalli commented on sqladmin/_queries.py:
```diff
@@ -230,6 +230,42 @@ async def _insert_async(self, data: dict[str, Any], request: Request) -> Any:
+                obj = self.model_view.model()
+                anyio.from_thread.run(
+                    self.model_view.on_model_change, row, obj, True, request
+                )
```
I believe this part wont work, as `obj` is not correct object. Please check `_insert_sync` and how they create obj and pass it to these functions

mmzeynalli commented on sqladmin/_queries.py:
```diff
@@ -230,6 +230,42 @@ async def _insert_async(self, data: dict[str, Any], request: Request) -> Any:
+            session.commit()
+            for obj, row in zip(objs, data):
+                anyio.from_thread.run(
+                    self.model_view.after_model_change, row, obj, True, request
```
same

mmzeynalli commented on sqladmin/_queries.py:
```diff
@@ -230,6 +230,42 @@ async def _insert_async(self, data: dict[str, Any], request: Request) -> Any:
+        objs = []
+        async with self.model_view.session_maker(expire_on_commit=False) as session:
+            for row in data:
+                obj = self.model_view.model()
```
same

mmzeynalli replied:
Also, do as you did above, user might want to change something between inserts in `after_model_change`

mmzeynalli commented on sqladmin/helpers.py:
```diff
@@ -182,6 +183,24 @@ def stream_to_csv(
+def parse_csv(
+    csv_content: bytes, columns: list[str], delimiter: str = ","
+) -> list[MultiDict]:
+    if csv_content[:3] == b"\xef\xbb\xbf":
+        csv_content = csv_content[3:]
```
?

vahidzhe replied:
Instead of manually stripping the BOM, we can just decode with the `utf-8-sig` codec, it removes the BOM if present and is a no-op otherwise:

```python
try:
    _csv_content = csv_content.decode("utf-8-sig").splitlines()
except UnicodeDecodeError as exc:
    raise ValueError("CSV file must be UTF-8 encoded.") from exc
```

This drops the extra `if`/slice and handles the same Excel-exported CSV case.

mmzeynalli commented on sqladmin/models.py:
```diff
@@ -1090,6 +1135,13 @@ async def delete_model(self, request: Request, pk: Any) -> None:
     async def insert_model(self, request: Request, data: dict) -> Any:
         return await Query(self).insert(data, request)
 
+    async def insert_many_models(
```
insert_models_many

mmzeynalli committed:
Merge branch 'main' into csv-import

mmzeynalli committed:
feat: Enhanced import logic

- Added `check_can_import` for more flexibility
- Added checkbox to skip on error
- Added progress bar for large imports
- Corresponding rows are shown if they are skipped, or aborted on error
- Added max file size enforcer

mmzeynalli committed:
fix: Fixed psql type problem

mmzeynalli committed:
feat: Enhance CSV import functionality

- Updated import process to support UTF-8 CSV files with NDJSON progress streaming.
- Introduced `import_max_rows` to limit the number of rows imported.
- Added validation for required columns and improved error handling for invalid data.
- Implemented hooks for custom row processing during import.
- Updated documentation to reflect new import options and usage examples.

mmzeynalli committed:
fix: Removed duplicate choose file button

vahidzhe commented on sqladmin/application.py:
```diff
@@ -756,6 +795,502 @@ async def export(self, request: Request) -> Response:
+                    return
+
+                processed += 1
+                form = Form(row)
```
Boolean values don't survive a round-trip. For non-nullable Boolean columns the scaffolded form uses a plain `BooleanField`, whose `false_values` is `(False, "false", "")`  so a CSV cell containing `False` (exactly what the exporter writes), or `0` / `no` / `off`, all validate to `True`. `helpers._coerce_bool` already implements the correct semantics; either reuse it for boolean columns before feeding the form, or use a `BooleanField` subclass with a wider `false_values` for the import form. An export -> import -> compare test would catch this.

vahidzhe commented on sqladmin/application.py:
```diff
@@ -756,6 +795,502 @@ async def export(self, request: Request) -> Response:
+                    model_view.model,
+                )
+
+                # Import forms may omit FK columns (when form_include_pk=False).
```
Columns kept from `row_data` here bypass type coercion entirely, so a FK column like `user_id` reaches `setattr()` as the raw string `"5"`. `_validate_foreign_key_values` already coerces the value via `coerce_column_value()` for the existence check, but the coerced value is discarded. On PostgreSQL/asyncpg this fails with a type error for any non-string column that isn't present in the form (FKs, PKs, ints, dates). Suggest coercing here , `coerce_column_value(mapper.columns[name], value)` , and turning a coercion failure into a row-level error instead of an unhandled exception.

mmzeynalli committed:
fix: coerce gaps
